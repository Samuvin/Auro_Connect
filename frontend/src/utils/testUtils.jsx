import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';

// Create a custom render function that includes providers
export const renderWithProviders = (ui, options = {}) => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
  });

  const AllTheProviders = ({ children }) => {
    return (
      <QueryClientProvider client={queryClient}>
        <BrowserRouter>
          {children}
        </BrowserRouter>
      </QueryClientProvider>
    );
  };

  return render(ui, { wrapper: AllTheProviders, ...options });
};

// Mock user data for testing
export const mockUser = {
  _id: '1',
  name: 'John Doe',
  username: 'johndoe',
  email: 'john@example.com',
  profilePicture: '/avatar.png',
  headline: 'Software Developer',
  connections: ['2', '3'],
  location: 'New York, NY'
};

// Mock post data for testing
export const mockPost = {
  _id: '1',
  content: 'This is a test post',
  author: mockUser,
  likes: ['2', '3'],
  comments: [
    {
      _id: '1',
      content: 'Great post!',
      user: mockUser,
      createdAt: new Date().toISOString()
    }
  ],
  image: 'https://example.com/image.jpg',
  createdAt: new Date().toISOString()
}; 