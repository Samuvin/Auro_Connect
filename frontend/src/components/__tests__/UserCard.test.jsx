import { screen, fireEvent } from '@testing-library/react';
import UserCard from '../UserCard';
import { renderWithProviders, mockUser } from '../../utils/testUtils';

describe('UserCard', () => {
  const defaultProps = {
    user: mockUser,
    isConnection: false
  };

  test('renders user information correctly', () => {
    renderWithProviders(<UserCard {...defaultProps} />);
    
    expect(screen.getByText('John Doe')).toBeInTheDocument();
    expect(screen.getByText('Software Developer')).toBeInTheDocument();
    expect(screen.getByText('2 connections')).toBeInTheDocument();
    expect(screen.getByAltText('John Doe')).toBeInTheDocument();
  });

  test('displays profile picture with fallback', () => {
    renderWithProviders(<UserCard {...defaultProps} />);
    
    const img = screen.getByAltText('John Doe');
    expect(img).toHaveAttribute('src', '/avatar.png');
  });

  test('displays fallback avatar when no profile picture', () => {
    const userWithoutPicture = { ...mockUser, profilePicture: null };
    renderWithProviders(
      <UserCard user={userWithoutPicture} isConnection={false} />
    );
    
    const img = screen.getByAltText('John Doe');
    expect(img).toHaveAttribute('src', '/avatar.png');
  });

  test('renders Connect button when not connected', () => {
    renderWithProviders(<UserCard {...defaultProps} />);
    
    expect(screen.getByText('Connect')).toBeInTheDocument();
  });

  test('renders Connected button when already connected', () => {
    renderWithProviders(
      <UserCard user={mockUser} isConnection={true} />
    );
    
    expect(screen.getByText('Connected')).toBeInTheDocument();
  });

  test('links to user profile', () => {
    renderWithProviders(<UserCard {...defaultProps} />);
    
    const profileLink = screen.getByRole('link');
    expect(profileLink).toHaveAttribute('href', '/profile/johndoe');
  });

  test('handles click on connect button', () => {
    renderWithProviders(<UserCard {...defaultProps} />);
    
    const connectButton = screen.getByText('Connect');
    fireEvent.click(connectButton);
    // In a real implementation, this would trigger a connection request
  });

  test('displays zero connections when user has no connections', () => {
    const userWithNoConnections = { ...mockUser, connections: [] };
    renderWithProviders(
      <UserCard user={userWithNoConnections} isConnection={false} />
    );
    
    expect(screen.getByText('0 connections')).toBeInTheDocument();
  });

  test('handles undefined connections array', () => {
    const userWithUndefinedConnections = { ...mockUser, connections: undefined };
    renderWithProviders(
      <UserCard user={userWithUndefinedConnections} isConnection={false} />
    );
    
    // The component shows " connections" when connections is undefined
    // Use a custom matcher since testing library normalizes whitespace
    expect(screen.getByText((content, element) => {
      return element?.textContent === ' connections';
    })).toBeInTheDocument();
  });

  // Snapshot Tests
  test('matches snapshot for default user card', () => {
    const { container } = renderWithProviders(<UserCard {...defaultProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('matches snapshot for connected user card', () => {
    const { container } = renderWithProviders(
      <UserCard user={mockUser} isConnection={true} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('matches snapshot for user with no connections', () => {
    const userWithNoConnections = { ...mockUser, connections: [] };
    const { container } = renderWithProviders(
      <UserCard user={userWithNoConnections} isConnection={false} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('matches snapshot for user with undefined connections', () => {
    const userWithUndefinedConnections = { ...mockUser, connections: undefined };
    const { container } = renderWithProviders(
      <UserCard user={userWithUndefinedConnections} isConnection={false} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
}); 