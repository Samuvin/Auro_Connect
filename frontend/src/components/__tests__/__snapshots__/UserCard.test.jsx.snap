// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`UserCard matches snapshot for connected user card 1`] = `
<div
  class="bg-white rounded-lg shadow p-4 flex flex-col items-center transition-all hover:shadow-md"
>
  <a
    class="flex flex-col items-center"
    href="/profile/johndoe"
  >
    <img
      alt="John Doe"
      class="w-24 h-24 rounded-full object-cover mb-4"
      src="/avatar.png"
    />
    <h3
      class="font-semibold text-lg text-center"
    >
      John Doe
    </h3>
  </a>
  <p
    class="text-gray-600 text-center"
  >
    Software Developer
  </p>
  <p
    class="text-sm text-gray-500 mt-2"
  >
    2
     connections
  </p>
  <button
    class="mt-4 bg-primary text-white px-4 py-2 rounded-md hover:bg-primary-dark transition-colors w-full"
  >
    Connected
  </button>
</div>
`;

exports[`UserCard matches snapshot for default user card 1`] = `
<div
  class="bg-white rounded-lg shadow p-4 flex flex-col items-center transition-all hover:shadow-md"
>
  <a
    class="flex flex-col items-center"
    href="/profile/johndoe"
  >
    <img
      alt="John Doe"
      class="w-24 h-24 rounded-full object-cover mb-4"
      src="/avatar.png"
    />
    <h3
      class="font-semibold text-lg text-center"
    >
      John Doe
    </h3>
  </a>
  <p
    class="text-gray-600 text-center"
  >
    Software Developer
  </p>
  <p
    class="text-sm text-gray-500 mt-2"
  >
    2
     connections
  </p>
  <button
    class="mt-4 bg-primary text-white px-4 py-2 rounded-md hover:bg-primary-dark transition-colors w-full"
  >
    Connect
  </button>
</div>
`;

exports[`UserCard matches snapshot for user with no connections 1`] = `
<div
  class="bg-white rounded-lg shadow p-4 flex flex-col items-center transition-all hover:shadow-md"
>
  <a
    class="flex flex-col items-center"
    href="/profile/johndoe"
  >
    <img
      alt="John Doe"
      class="w-24 h-24 rounded-full object-cover mb-4"
      src="/avatar.png"
    />
    <h3
      class="font-semibold text-lg text-center"
    >
      John Doe
    </h3>
  </a>
  <p
    class="text-gray-600 text-center"
  >
    Software Developer
  </p>
  <p
    class="text-sm text-gray-500 mt-2"
  >
    0
     connections
  </p>
  <button
    class="mt-4 bg-primary text-white px-4 py-2 rounded-md hover:bg-primary-dark transition-colors w-full"
  >
    Connect
  </button>
</div>
`;

exports[`UserCard matches snapshot for user with undefined connections 1`] = `
<div
  class="bg-white rounded-lg shadow p-4 flex flex-col items-center transition-all hover:shadow-md"
>
  <a
    class="flex flex-col items-center"
    href="/profile/johndoe"
  >
    <img
      alt="John Doe"
      class="w-24 h-24 rounded-full object-cover mb-4"
      src="/avatar.png"
    />
    <h3
      class="font-semibold text-lg text-center"
    >
      John Doe
    </h3>
  </a>
  <p
    class="text-gray-600 text-center"
  >
    Software Developer
  </p>
  <p
    class="text-sm text-gray-500 mt-2"
  >
     connections
  </p>
  <button
    class="mt-4 bg-primary text-white px-4 py-2 rounded-md hover:bg-primary-dark transition-colors w-full"
  >
    Connect
  </button>
</div>
`;
